src/CleverDict.jl

```diff
@@ -0,0 +1,157 @@
+"""
+    CleverDict{K, V}
+
+A smart storage type for managing sequential objects with non-decreasing integer
+indices.
+
+Provided no keys are deleted, the backing storage is a `Vector{V}`. Once a key
+has been deleted, the backing storage switches to a standard Julia `Dict{K,
+V}`.
+
+The i'th ordered element can be obtained with `c[LinearIndex(i)]`.
+
+Use `new_key(c::CleverDict)` to obtain the next key in the sequential order.
+
+When `empty!` is called, the largest index is stored in `offset` so that new
+keys are generated from `offset + 1`.
+
+Overload the functions `index_to_key` and `key_to_index` to enable mappings
+between the integer index of the vector (plus the offset) and the dictionary key.
+"""
+mutable struct CleverDict{K, V}
+    last_index::Int
+    vector::Vector{V}
+    offset::Int
+    dict::Union{Nothing, Dict{K, V}}
+    CleverDict{K, V}() where {K, V} = new{K, V}(0, V[], 0, nothing)
+end
+
+"""
+    index_to_key(::Type{K}, index::Int)
+
+Create a new key associated with the integer value `index`.
+"""
+function index_to_key(::Type{K}, index::Int) where {K}
+    error("You need to define `index_to_key`.")
+end
+
+"""
+    key_to_index(key::K)
+
+Map `key` to an integer valued index.
+"""
+function key_to_index(key::K) where {K}
+    error("You need to define `key_to_index`.")
+end
+
+"""
+    new_key(c::CleverDict{K, V}) where {K, V}
+
+Return the next available key in the ordered sequence.
+"""
+function new_key(c::CleverDict{K, V}) where {K, V}
+    c.last_index += 1
+    return index_to_key(K, c.last_index)
+end
+
+function Base.empty!(c::CleverDict{K, V}) where {K, V}
+    empty!(c.vector)
+    c.offset = c.last_index
+    c.dict = nothing
+    return
+end
+
+function Base.getindex(c::CleverDict{K, V}, key::K) where {K, V}
+    if c.dict === nothing
+        # Case I) no call to `MOI.delete!`, so return the offsetted element:
+        return c.vector[key_to_index(key) - c.offset]
+    end
+    # Case II) `MOI.delete!` must have been called, so return the element
+    #          from the dictionary.
+    return c.dict[key]
+end
+
+function Base.setindex!(c::CleverDict{K, V}, val::V, key::K) where {K, V}
+    push!(c.vector, val)
+    if c.dict !== nothing
+        c.dict[key] = val
+    end
+    return val
+end
+
+struct LinearIndex
+    i::Int
+end
+Base.getindex(c::CleverDict, index::LinearIndex) = c.vector[index.i]
+
+function Base.delete!(c::CleverDict{K, V}, key::K) where {K, V}
+    if c.dict === nothing
+        c.dict = Dict{K, V}()
+        for (i, info) in enumerate(c.vector)
+            if key.value == i + c.offset
+                continue
+            end
+            c.dict[index_to_key(K, i + c.offset)] = info
+        end
+        splice!(c.vector, key.value - c.offset)
+    else
+        for (i, v) in enumerate(c.vector)
+            if v == key
+                splice!(c.vector, i)
```

Comment:
This makes deletions `O(n)`. Is it feasible to set the `vector` to nothing if the dict is in use instead of keeping it updated?

Reply:
There is a trade-off:

 - making delete `O(n)`, and maintaining an ordered list of constraint indices
 - setting `vector` to `nothing`, and building an ordered list of variables on demand for calls like `get(model, ::ObjectiveFunction` or `get(model, ::ConstraintFunction)`, which needs a map between the Gurobi column and the MOI index.

Which would you prefer?

Reply:
What about using `OrderedDict` instead of `Dict`?

Reply:
So we could use an `OrderedDict`, but we would still need to build the list of variables on demand, because there isn't a mechanism to get the nth element of an OrderedDict (https://github.com/JuliaCollections/OrderedCollections.jl/blob/master/src/ordered_dict.jl).